### PR TITLE
Identity Deletion Grace Period

### PR DESCRIPTION
**Pull request by JLSRKonk**
# Readiness checklist

-   [x] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

### CODE REVIEW

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        var identities = await _identitiesRepository.FindAllWithApprovedDeletionProcess(cancellationToken, track: true);
+
+        foreach (var identity in identities)
+        {
+            var deletionProcess = identity.DeletionProcesses.First(d => d.IsActive());
+
+            if (deletionProcess.GracePeriodReminder1SentAt == null)
+            {
+                var daysToDeletion = (deletionProcess.GracePeriodEndsAt! - SystemTime.UtcNow).Value.Days;
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs:30]: I see this exact line of code at least 3 times.

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    await _pushSenderService.SendNotification(identity.Address, new DeletionProcessGracePeriodNotification(daysToDeletion), cancellationToken);
+                    identity.DeletionGracePeriodReminder3Sent();
+
+                    await _identitiesRepository.Update(identity, cancellationToken);
+                }
+            }
+        }
+
+        return new DeletionProcessGracePeriodResponse();
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs:68]: What's the point in having this response?

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs [hunk 3]
```diff
@@ -76,6 +76,24 @@ public IdentityDeletionProcess StartDeletionProcessAsOwner(DeviceId asDevice)
         return deletionProcess;
     }
 
+    public void DeletionGracePeriodReminder1Sent()
+    {
+        var deletionProcess = DeletionProcesses.First(d => d.IsActive());
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:81]: What if we don't have an active deletion process?
This is a public method, so you can't assume it will only be called by your handler where you know there's an active process.

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs [hunk 4]
```diff
@@ -76,6 +76,24 @@ public IdentityDeletionProcess StartDeletionProcessAsOwner(DeviceId asDevice)
...
+        deletionProcess.GracePeriodReminder1Sent(Address);
+    }
+
+    public void DeletionGracePeriodReminder2Sent()
+    {
+        var deletionProcess = DeletionProcesses.First(d => d.IsActive());
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:87]: Same as above.

File: Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs [hunk 5]
```diff
@@ -76,6 +76,24 @@ public IdentityDeletionProcess StartDeletionProcessAsOwner(DeviceId asDevice)
...
+        deletionProcess.GracePeriodReminder2Sent(Address);
+    }
+
+    public void DeletionGracePeriodReminder3Sent()
+    {
+        var deletionProcess = DeletionProcesses.First(d => d.IsActive());
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/Identity.cs:93]: Same as above.

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    public async Task<DeletionProcessGracePeriodResponse> Handle(DeletionProcessGracePeriodCommand request, CancellationToken cancellationToken)
+    {
+        var identities = await _identitiesRepository.FindAllWithApprovedDeletionProcess(cancellationToken, track: true);
+
+        foreach (var identity in identities)
+        {
+            var deletionProcess = identity.DeletionProcesses.First(d => d.IsActive());
+
+            if (deletionProcess.GracePeriodReminder1SentAt == null)
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs:28]: Shouldn't we validate which reminders we are going to send?
Otherwise we might end up sending 3 reminders at once (obviously only some specific scenarios e.g., configuration changes)

File: Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionConfiguration.cs [hunk 7]
```diff
@@ -2,4 +2,26 @@
 public static class IdentityDeletionConfiguration
 {
     public static int LengthOfGracePeriod { get; set; } = 30;
+
+    public static GracePeriodNotification GracePeriodNotification1 { get; set; } = new()
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Domain/Entities/Identities/IdentityDeletionConfiguration.cs:6]: Should probably replace these properties with their actual intended values.

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 8]
```diff
@@ -170,25 +170,6 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                     b.ToTable("AspNetUsers", (string)null);
                 });
 
-            modelBuilder.Entity("Backbone.Modules.Devices.Domain.Entities.Challenge", b =>
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs:173]: Why is this change here?

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        A.CallTo(() => mockIdentitiesRepository.FindAllWithApprovedDeletionProcess(A<CancellationToken>._, A<bool>._))
+            .MustHaveHappenedOnceExactly();
+        A.CallTo(() => mockIdentitiesRepository.Update(A<Identity>._, A<CancellationToken>._))
+            .MustNotHaveHappened();
+        A.CallTo(() => mockPushNotificationSender.SendNotification(A<IdentityAddress>._, A<object>._, A<CancellationToken>._))
+            .MustNotHaveHappened();
+    }
+
+    [Fact]
+    public async Task Sends_First_Reminder()
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs:40]: Shouldn't this also validate that only the first one is sent?

File: Modules/Devices/test/Devices.Domain.Tests/Identities/DeletionProcessGracePeriodTests.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
+﻿using Backbone.DevelopmentKit.Identity.ValueObjects;
+using Backbone.Modules.Devices.Domain.Aggregates.Tier;
+using Backbone.Modules.Devices.Domain.Entities.Identities;
+using Backbone.Modules.Devices.Domain.Tests.Identities.TestDoubles;
+using Backbone.Tooling;
+using FluentAssertions;
+using Xunit;
+
+namespace Backbone.Modules.Devices.Domain.Tests.Identities;
+public class DeletionProcessGracePeriodTests : IDisposable
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Devices/test/Devices.Domain.Tests/Identities/DeletionProcessGracePeriodTests.cs:10]: Need tests after adding the missing validations to these methods (comments above in the Identity class).

**Review by daniel-almeida-konkconsulting — CHANGES_REQUESTED** [review]

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs (hunk 6 above)
Comment (JLSRKonk) [reply Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs:28]: Is it expected to only be able to send 1 notification per run of the handler for a given process?
So, in a scenario where for some reason we couldn't deliver the first two notifications in time, should only the last one be sent and the other ones discarded?

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 8 above)
Comment (JLSRKonk) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs:173]: I'm not actually sure where this is coming from.
The same thing happened when I merged the changes from main into the identity-deletion and recreated the migrations.

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs (hunk 6 above)
Comment (tnotheis) [reply Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs:28]: In the long run: yes. But for now, as discussed in chat, we can send all of the missing notifications, if you don't find an easy way to avoid it.

File: Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 8 above)
Comment (JLSRKonk) [reply Modules/Devices/src/Devices.Infrastructure.Database.SqlServer/Migrations/ApplicationDbContextModelSnapshot.cs]: The migrations should be fixed now.

File: Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs (hunk 6 above)
Comment (JLSRKonk) [reply Modules/Devices/src/Devices.Application/Identities/Commands/DeletionProcessGracePeriod/Handler.cs]: I changed it so now it only sends the latest required notification of each process per run of the handler.

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            .Returns(new List<Identity>());
+
+        var handler = CreateHandler(mockIdentitiesRepository, mockPushNotificationSender);
+
+        // Act
+        await handler.Handle(new DeletionProcessGracePeriodCommand(), CancellationToken.None);
+
+        // Assert
+        A.CallTo(() => mockIdentitiesRepository.FindAllWithApprovedDeletionProcess(A<CancellationToken>._, A<bool>._))
+            .MustHaveHappenedOnceExactly();
```
Comment (tnotheis) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs:32]: This does not have to have happened. It doesn't matter how the Handler gets the data it needs.

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+using Backbone.Tooling;
+using FakeItEasy;
+using FluentAssertions;
+using Xunit;
+
+namespace Backbone.Modules.Devices.Application.Tests.Tests.Identities.Commands.DeletionProcessGracePeriod;
+public class HandlerTests
+{
+    [Fact]
+    public async Task No_Active_Deletion_Processes()
```
Comment (tnotheis) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs:16]: This name doesn't match the implementation.
```suggestion
    public async Task No_identities_with_an_approved_deletion_process_exist()
```

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        // Arrange
+        var beginProcessDate = DateTime.Parse("2000-01-01");
+        SystemTime.Set(beginProcessDate);
+
+        var activeIdentity = TestDataGenerator.CreateIdentityWithOneDevice();
+        var deletionProcess = activeIdentity.StartDeletionProcessAsOwner(activeIdentity.Devices[0].Id);
+
+        var reminderSentDate = deletionProcess.GracePeriodEndsAt!.Value
+            .AddDays(-IdentityDeletionConfiguration.GracePeriodNotification1.Time);
```
Comment (tnotheis) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs:50]: Here you reimplement the logic of how the date is calculated. That's not good. 
Hardcode the value instead. See the other tests as well

File: Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        A.CallTo(() => mockPushNotificationSender.SendNotification(A<IdentityAddress>._, A<object>._, A<CancellationToken>._))
+            .MustNotHaveHappened();
+    }
+
+    [Fact]
+    public async Task Sends_First_Reminder()
+    {
+        // Arrange
+        var beginProcessDate = DateTime.Parse("2000-01-01");
+        SystemTime.Set(beginProcessDate);
```
Comment (tnotheis) [inline Modules/Devices/test/Devices.Application.Tests/Tests/Identities/Commands/DeletionProcessGracePeriod/HandlerTests.cs:44]: What's that code for?
1. You never use the `beginProcessDate`.
2. You don't care what the processe's `CreatedAt` date is.

Remove it for the other tests as well

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — CHANGES_REQUESTED** [review]

**Review by tnotheis — APPROVED** [review]
